smarts/env/wrappers/record_video.py

```diff
@@ -0,0 +1,166 @@
+
+        self.episode_trigger = episode_trigger
+        self.step_trigger = step_trigger
+        self.video_recorder = None
+
+        self.video_folder = os.path.abspath(video_folder)
+        # Create output folder if needed
+        if os.path.isdir(self.video_folder):
+            logger.warn(
+                f"Overwriting existing videos at {self.video_folder} folder (try specifying a different `video_folder` for the `RecordVideo` wrapper if this is not desired)"
+            )
+        os.makedirs(self.video_folder, exist_ok=True)
+
+        self.name_prefix = name_prefix
+        self.step_id = 0
+        self.video_length = video_length
+
+        self.recording = False
+        self.recorded_frames = 0
+        self.is_vector_env = getattr(env, "is_vector_env", False)
+        self.episode_id = 0
+
+    def reset(self, **kwargs):
+        """Reset."""
+        observations = super().reset(**kwargs)
+        if not self.recording and self._video_enabled():
+            self.start_video_recorder()
+        return observations
+
+    def start_video_recorder(self):
+        """Start video."""
+        self.close_video_recorder()
+
+        video_name = f"{self.name_prefix}-step-{self.step_id}"
+        if self.episode_trigger:
+            video_name = f"{self.name_prefix}-episode-{self.episode_id}"
+
+        base_path = os.path.join(self.video_folder, video_name)
+        self.video_recorder = video_recorder.VideoRecorder(
+            env=self.env,
+            base_path=base_path,
+            metadata={"step_id": self.step_id, "episode_id": self.episode_id},
+        )
+
+        self.video_recorder.capture_frame()
+        self.recorded_frames = 1
+        self.recording = True
+
+    def _video_enabled(self):
+        if self.step_trigger:
+            return self.step_trigger(self.step_id)
+        else:
+            return self.episode_trigger(self.episode_id)
+
+    def step(self, action):
+        """Step."""
+        observations, rewards, dones, infos = super().step(action)
+
+        # increment steps and episodes
+        self.step_id += 1
+        if not self.is_vector_env:
+            if dones:
+                self.episode_id += 1
+        elif dones[0]:
+            self.episode_id += 1
+
+        if self.recording:
+            self.video_recorder.capture_frame()
+            self.recorded_frames += 1
+            if self.video_length > 0:
+                if self.recorded_frames > self.video_length:
+                    self.close_video_recorder()
+            else:
+                if not self.is_vector_env:
+                    if dones:
+                        self.close_video_recorder()
+                elif dones[0]:
+                    self.close_video_recorder()
+
+        elif self._video_enabled():
+            self.start_video_recorder()
+
+        return observations, rewards, dones, infos
+
+    def close_video_recorder(self) -> None:
```

Comment:
... but no `-> None` here.  (`None` is not really a _type_!)

Reply:
`None` is singleton and is allowed as a type as an exception to the rule (alongside `...`.) (although Python 3.10 has `NoneType`)